packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js

```diff
@@ -3,277 +3,646 @@ import { FreehandLine as FreehandLineComponent } from '@plugins/drawingTools/com
+        if (spliceSpan > self.points.length / 2) {
+          spliceSpan = Math.abs(dragPointIndex - closePointIndex)
+          spliceReverse = dragPointIndex < closePointIndex
+        }
+
+        if (spliceReverse && dragPointIndex < spliceSpan) {
+          spliceThroughBeginning = true;
+        } else if (!spliceReverse && (dragPointIndex + spliceSpan) > self.points.length) {
+          spliceThroughBeginning = true;
+        }
+      }
+
+      self.drawActions.push({
+        type: 'splice-close-point',
+        pointIndexDrag: dragPointIndex,
+        pointIndexClose: closePointIndex,
+        pointsFromDrag: [],
+        pointsFromClose: [],
+        pointsSpliceCount: 0,
+        isReversed: spliceReverse,
+        throughBeginning: spliceThroughBeginning
+      });
+
+      self.splicePathSetup()
+      self.redoActionsClear()
+    },
+
+    splicePathSetup() {
+      // for purposes of undo/redo we use the action to set these properties
+      let action = self.drawActions.at(-1)
+      self.spliceClosePointIndex = action.pointIndexClose
+      self.spliceReverse = action.isReversed
+      self.spliceThroughBeginning = action.throughBeginning
+      self.spliceActive = true
+      self.setClosePoint(self.points[action.pointIndexClose])
+    },
+
+    spliceReset() {
+      self.splicePoints.length = 0
+      self.spliceActive = false
+      self.spliceReverse = false
+      self.spliceThroughBeginning = false
+      self.spliceDragPointIndex = null
+      self.spliceClosePointIndex = null
     },
 
-    setTargetPoint(point) {
-      self.targetPoint = point ? self.selectPoint(point) : null
+    move() {
+      // do nothing... tool required but not used in this context
     },
 
-    shortenPath() {
-      let lengthToRemove = 20
-      while (lengthToRemove--) {
-        self.pathX.pop()
-        self.pathY.pop()
+    inactive() {
+      while (['splice-drag-point', 'splice-close-point', 'splice-append'].indexOf(self.drawActions.at(-1).type) > -1) {
+        self.undo()
+        self.redoActionsClear()
       }
     },
 
-    splice(startIndex, endIndex) {
-      let firstPoint = Math.min(startIndex, endIndex)
-      let lastPoint = Math.max(startIndex, endIndex)
-      /*
-        Reverse the path, if necessary, so that the first point clicked
-        is always the draggable point.
-      */
-      if (firstPoint === endIndex) {
-        const lastIndex = self.points.length - 1
-        self.pathX.reverse()
-        self.pathY.reverse()
-        firstPoint = lastIndex - startIndex
-        lastPoint = lastIndex - endIndex
+    finish() {
+      if (self.points.length < self.minimumPoints) {
+        return self.tool.deleteMark(self)
+      }
+
+      if (self.isDragging) {
+        self.appendPathEnd()
+        self.finished = false
+      } else {
+        self.pathX = []
+        self.pathY = []
+        self.points.forEach(point => {
+          self.pathX.push(point.x)
+          self.pathY.push(point.y)
+        })
```

Comment:
@eatyourgreens - made my way back around to this... I haven't figured out how the mark gets pathX and pathY initially.. Where in the code would that occur? If we're fetching from Caesar, it's going to be separate, right? Or is this some aspect of data that would be stored in Panoptes?

Reply:
When you first create a mark, for any drawing tool, you pass in a snapshot. For the freehand line, that would look like:

```js
FreehandLine.create({ pathX, pathY })
```

where `pathX` and `pathY` are the initial values for the x and y coords. 

Editing a mark updates its model. The classification holds references to all the current mark models. When we save to Panoptes, we take a final snapshot of those models and send that back in `classification.annotations`.

Reply:
Thanks for the answer but it doesn't quite get at what I'm wanting to know... I'll try a more nuanced way of asking the question:

Where does the original pathX / pathY come from if its going to come from somewhere? Where in the code does it travel/flow? My assumption is that there is a Panoptes API call to get a subject and then it gets initiated / stored somewhere and then it must get pulled out into the code at some point. So far I have a rough understanding of the inheritance model as follows: SubjectViewer > SingleImageViewerContainer > SingleImageViewer > InteractionLayerContainer > InteractionLayer > DrawingToolMarks. There's a lot of moving pieces and interdependencies I don't fully grok yet but my hunch is somewhere in that chain I'll understand why it's pathX / pathY and not some other data structure.

Another way to paraphrase, when we get to the FreehandLineTool.createMark() function what is the chain of events beforehand from a file/function-to-file/function way?

Reply:
NB I'm suggesting `Array.map` here because looping with `Array.push` is slow.

Reply:
Added the ability to handle both `pathX/pathY` and `points` in `createMark()`. Also removed all `push()` and replaced with appropriate methods.

Reply:
Bumping this because it's still using a loop and pushing points one at a time. Mapping `self.points` to `self.pathX` and `self.pathY` does the same thing in fewer operations.